pkg/topology/kademlia/kademlia.go

```diff
@@ -883,22 +883,24 @@ func (k *Kad) Connected(ctx context.Context, peer p2p.Peer, forceConnection bool
 				return err
 			}
 			_ = k.p2p.Disconnect(randPeer, "kicking out random peer to accommodate node")
-			return k.onConnected(ctx, address)
+			return k.onConnected(ctx, address, reachable)
 		}
 		if !forceConnection {
 			return topology.ErrOversaturated
 		}
 	}
 
-	return k.onConnected(ctx, address)
+	return k.onConnected(ctx, address, reachable)
 }
 
-func (k *Kad) onConnected(ctx context.Context, addr swarm.Address) error {
+func (k *Kad) onConnected(ctx context.Context, addr swarm.Address, reachable bool) error {
 	if err := k.Announce(ctx, addr, true); err != nil {
 		return err
 	}
 
-	k.knownPeers.Add(addr)
+	if reachable {
+		k.knownPeers.Add(addr)
```

Comment:
not sure about this as it changes the semantics of `knownPeers`. so it is no longer `knownPeers` but `knownAndReachable` peers. this will make connectivity issues more difficult to detect using the debug apis

Reply:
I'm ok with removing this change. It was added to deal with the case that if the unreachable peer connects, then some time later disconnects, kademlia could possibly attempt to reconnect to it. But this is not a common case and we can let kademlia fail to dial to this peer, it's not a big issue.